Add the 'initializes' attribute langref and support

haopliu opened the pull request:
We propose adding a new LLVM attribute, `initializes((Lo1,Hi1),(Lo2,Hi2),...)`, which expresses the notion of memory space (i.e., intervals, in bytes) that the argument pointing to is initialized in the function.

Will commit the attribute inferring in the follow-up PRs.

https://discourse.llvm.org/t/rfc-llvm-new-initialized-parameter-attribute-for-improved-interprocedural-dse/77337

haopliu committed:
Add 'initialized' attribute langref

jvoung commented on llvm/docs/LangRef.rst:
```diff
@@ -1621,6 +1621,28 @@ Currently, only the following parameter attributes are defined:
 
+``initialized((Lo1,Hi1),...)``
+    This attribute is a list of const ranges in ascending order with no
+    overlapping or continuous. It indicates that the function initializes the
```
nit: "with no overlapping or adjoining list elements"? or something like that (felt like it was missing a word at the end at least)

aeubanks replied:
We should start this off with a one sentence overview of the attribute. So starting with something like "This indicates that the function initializes the ranges of the pointer parameter's memory." Then describe what "initialize" means. Then the random details like non-overlapping/continuous ranges at the end.

jvoung commented on llvm/docs/LangRef.rst:
```diff
@@ -1621,6 +1621,28 @@ Currently, only the following parameter attributes are defined:
+    and no special accesses (such as volatile access or untrackable capture)
+    before the initialization in the function. LoN/HiN are 64-bit ints;
+    negative values are allowed in case a pointer to partway through the
+    allocation is passed to.
```
nit: "in case the argument points partway into an allocation." ?

haopliu replied:
Done!

jvoung commented on llvm/docs/LangRef.rst:
```diff
@@ -1621,6 +1621,28 @@ Currently, only the following parameter attributes are defined:
+    as through other arguments.
+
+    The ``writable`` or ``dereferenceable`` attribute does not imply
+    ``initialized`` attribute, and ``initialized`` does not imply ``writeonly``
```
nit: "does not imply [the] ``initialized`` attribute, and ... ``writeonly``, since ``initialized`` allows reading from the pointer after writing." ?

haopliu replied:
Done!

aeubanks commented on llvm/docs/LangRef.rst:
```diff
@@ -1621,6 +1621,28 @@ Currently, only the following parameter attributes are defined:
+    since cases that read from the pointer after write, can be ``initialized``
+    but not ``writeonly``.
+
+    Note that this attribute does not apply to the unwind edge: the memory may
```
as mentioned in the RFC, we should be more accurate here. the part of the attribute where the memory is read from before a write still must apply on the unwind edge. Also, I'd move this up since this is important to the semantics

aeubanks replied:
sorry, I meant that the memory is *not* read from before a write

haopliu replied:
Done!

haopliu committed:
Update the LangRef

haopliu committed:
Refactor the argument uses collecting to a function

aeubanks commented:
unintentional add of FunctionAttrs.cpp change?

> we should combine this PR with the PR that adds support for this in LLVM, or else it's weird if we're documenting something that LLVM doesn't support yet

specifically just the attribute support (e.g. bitcode) not any of the inference or usage of it

aeubanks commented on llvm/docs/LangRef.rst:
```diff
@@ -1621,6 +1621,31 @@ Currently, only the following parameter attributes are defined:
+    written to when unwinding happens.
+
+    The ``writable`` or ``dereferenceable`` attribute does not imply the
+    ``initialized`` attribute, and ``initialized`` does not imply ``writeonly``
```
I'd separate this into two sentences.

haopliu replied:
Done!

aeubanks commented on llvm/docs/LangRef.rst:
```diff
@@ -1621,6 +1621,31 @@ Currently, only the following parameter attributes are defined:
+    ``initialized`` attribute, and ``initialized`` does not imply ``writeonly``
+    since ``initialized`` allows reading from the pointer after writing.
+
+    This attribute is a list of const ranges in ascending order with no
```
s/const/constant

aeubanks commented on llvm/docs/LangRef.rst:
```diff
@@ -1621,6 +1621,31 @@ Currently, only the following parameter attributes are defined:
+    since ``initialized`` allows reading from the pointer after writing.
+
+    This attribute is a list of const ranges in ascending order with no
+    overlapping or adjoining list elements. LoN/HiN are 64-bit ints, and
```
I haven't heard "adjoining" used in this context, I think "consecutive" is more commonly used.

aeubanks replied:
put `LoN`/`HiN` in double backticks

nikic commented:
As a heads up, the inference implementation for this (or at least the available prototype) does not look viable from a compile-time perspective: https://llvm-compile-time-tracker.com/compare.php?from=69b09b43b0a2057918078edb401adab888d1014b&to=0bd68ae2d56783377acc0aa5d7958b47411b8342&stat=instructions:u

haopliu committed:
Add 'initialized' attribute support and undo the FunctionAttrs refactoring.

haopliu committed:
Update the LangRef (round 2)

haopliu commented on llvm/docs/LangRef.rst:
```diff
@@ -1621,6 +1621,31 @@ Currently, only the following parameter attributes are defined:
     ``argmem: write``.
 
+``initialized((Lo1,Hi1),...)``
+    This attribute indicates that the function initializes the ranges of the
```
Done!

haopliu commented on llvm/docs/LangRef.rst:
```diff
@@ -1621,6 +1621,31 @@ Currently, only the following parameter attributes are defined:
+    special accesses (such as volatile access or untrackable capture) before
+    the initialization write in the function.
+
+    This attribute implies that the function initializes and does not read
```
Done!

haopliu commented:
> As a heads up, the inference implementation for this (or at least the available prototype) does not look viable from a compile-time perspective: https://llvm-compile-time-tracker.com/compare.php?from=69b09b43b0a2057918078edb401adab888d1014b&to=0bd68ae2d56783377acc0aa5d7958b47411b8342&stat=instructions:u

Thanks for this heads up! This PR only adds the attribute support. Will further tune the infering performance and post the updated compile time tracker in [the inference PR](https://github.com/llvm/llvm-project/pull/84869).

kuhar commented on llvm/include/llvm/ADT/FoldingSet.h:
```diff
@@ -362,6 +362,21 @@ class FoldingSetNodeID {
+    if (Size == 0)
+      return;
+
+    for (const auto &[Start, End] : Ranges) {
```
nit: const & doesn't do anything here
```suggestion
    for (auto [Start, End] : Ranges) {
```

kuhar commented on llvm/include/llvm/ADT/FoldingSet.h:
```diff
@@ -362,6 +362,21 @@ class FoldingSetNodeID {
   }
 
+  void AddRanges(const SmallVector<std::pair<int64_t, int64_t>, 16> &Ranges) {
+    unsigned Size = Ranges.size();
```
size_t?

haopliu replied:
Done, thanks!

haopliu committed:
Update FoldingSet.h

jvoung commented on llvm/include/llvm/IR/Attributes.h:
```diff
@@ -226,6 +236,10 @@ class Attribute {
 
+  /// Return the attribute's value as a const range list. This requires the
+  /// attribute to be a const range list attribute.
+  SmallVector<std::pair<int64_t, int64_t>, 16> getValueAsRanges() const;
```
Is it possible to return an ArrayRef instead of a copy? Will the underlying storage lifetime work out?

jvoung commented on llvm/lib/AsmParser/LLParser.cpp:
```diff
@@ -2364,6 +2381,54 @@ bool LLParser::parseOptionalDerefAttrBytes(lltok::Kind AttrKind,
+  if (parseInt64(Range.first))
+    return true;
+
+  if (EatIfPresent(lltok::comma)) {
```
should it be an error to be missing the comma?

jvoung commented on llvm/lib/Bitcode/Writer/BitcodeWriter.cpp:
```diff
@@ -930,11 +932,24 @@ void ModuleBitcodeWriter::writeAttributeGroupTable() {
           Record.push_back(VE.getTypeID(Attr.getValueAsType()));
-      } else {
+      } else if (Attr.isConstantRangeAttribute()) {
         assert(Attr.isConstantRangeAttribute());
```
nit: can remove the assert?

jvoung commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -2044,6 +2052,19 @@ void Verifier::verifyParameterAttrs(AttributeSet Attrs, Type *Ty,
+    Check(!Inits.empty(), "Attribute 'initialized' does not support empty list",
+          V);
+
+    for (size_t i = 1; i < Inits.size(); i++) {
```
perhaps check low < high for the same "i"?

aeubanks commented on llvm/include/llvm/IR/Attributes.h:
```diff
@@ -186,6 +193,9 @@ class Attribute {
   bool isTypeAttribute() const;
 
+  /// Return true if the attribute is a const range list attribute.
+  bool isConstRangeListAttribute() const;
```
+1, no need to abbreviate

aeubanks commented on llvm/docs/LangRef.rst:
```diff
@@ -1621,6 +1621,31 @@ Currently, only the following parameter attributes are defined:
+    parameter. Other arbitrary accesses to the same memory via other pointers
+    are allowed.
+
+    Note that this attribute does not apply to the unwind edge: the
```
this is now redundant with the unwind comment above

nikic commented:
> > As a heads up, the inference implementation for this (or at least the available prototype) does not look viable from a compile-time perspective: https://llvm-compile-time-tracker.com/compare.php?from=69b09b43b0a2057918078edb401adab888d1014b&to=0bd68ae2d56783377acc0aa5d7958b47411b8342&stat=instructions:u
> 
> Thanks for this heads up! This PR only adds the attribute support. Will further tune the inferring performance and post the updated compile time tracker in [the inference PR](https://github.com/llvm/llvm-project/pull/84869).

The main thing I would be concerned about at this point is whether the entire approach is viable or not -- I'd rather not add the attribute (plus a whole new attribute *kind*) if it later turns out that this is too expensive.

-----

Anyway, a couple of high level comments:

 * I'm not sure this is the ideal attribute name. `initialized` to me sounds like this attribute promises that the memory is initialized on entry to the function, not that the function will initialize it. Maybe `will_initialize` is a better name? Or `initializes`?
 * How important is it to represent multiple ranges? We have *just* added support for ConstantRange attributes, so a single range can be represented out of the box. This also avoids the tricky question of how to deal with non-zero start offsets in DSE, which we don't really have AA support for right now. My intuition would be that you usually get whole object initializations -- but maybe this is not the case due to padding holes?
 * Failing that, I feel like we should still piggy-back more on the new ConstantRange attribute kind. We could convert that into a list of ConstantRange stored as TrailingObjects and make the current single-range case a special case of the general multiple-range case.

aeubanks commented:
> The main thing I would be concerned about at this point is whether the entire approach is viable or not -- I'd rather not add the attribute (plus a whole new attribute _kind_) if it later turns out that this is too expensive.
+1
> 
> Anyway, a couple of high level comments:
> 
> * I'm not sure this is the ideal attribute name. `initialized` to me sounds like this attribute promises that the memory is initialized on entry to the function, not that the function will initialize it. Maybe `will_initialize` is a better name? Or `initializes`?
Makes sense, `initializes` sounds good.
> * How important is it to represent multiple ranges? We have _just_ added support for ConstantRange attributes, so a single range can be represented out of the box. This also avoids the tricky question of how to deal with non-zero start offsets in DSE, which we don't really have AA support for right now. My intuition would be that you usually get whole object initializations -- but maybe this is not the case due to padding holes?
Yes the concern was padding holes. @haopliu do you have any data on how often padding prevents whole object initialization?

aeubanks commented:
Even if we represent padding more precisely with multiple ranges in `initializes`, clang generates a `memset` of the entire alloca and DSE can only remove the part of the memset that corresponds to the first range and the last range, since it doesn't split memsets. So if we had one padding hole and used multiple ranges to represent everything else, we'd still end up with a store to the padding hole.

I think [`!tbaa.struct`](https://llvm.org/docs/LangRef.html#tbaa-struct-metadata) was designed for this, but [we don't emit it for `-ftrivial-auto-var-init`](https://godbolt.org/z/x1WWc5de8). Also it might not apply to `memset` right now, only `memcpy`?

aeubanks commented:
godbolt seems to not show metadata (I'm not seeing it on a `memcpy` on godbolt but am seeing it locally), but my point still stands

aeubanks commented:
actually I'm unsure if it would be legal to remove stores on padding bytes

haopliu committed:
Change the attribute type to ConstantRangeList

haopliu committed:
Rename attr to initializes

haopliu commented:
Renamed the attr to "initializes" and changed the attr type to ConstantRangeList.

For the compilation-time regression, we further optimized the attr inference and the new compilation cost looks good:
https://llvm-compile-time-tracker.com/compare.php?from=762cbd82da4debf8b026a4eb4ade66720acf3182&to=f2f31b0407f20e4000c5bb290215ae6614a28837&stat=instructions%3Au

As for multiple ranges, it seem not uncommon, at least in our internal benchmarks. Here is the data we got:
- The range size: 1-23, 25-31, 33-34, 37, 39, 41, 42, 45, 46, 49, 50, 54, 55, 59, 75, 76, 114, 216
- Percent: #. attrs whose size = 1 VS. > 1 is about 3 : 1.
- It could be caused by padding or a function just initializes the specific non-consecutive fields.

Meanwhile, we would like to have a good extensibility with multiple ranges. We experimented a prototype that improves DSE with zero start offsets/attributes and the result looks promising. We are exploring how to deal with non-zero start offsets in DSE :-D

Please take another look at your convenience. Thank you all!

jvoung commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -2032,6 +2033,30 @@ void Verifier::verifyParameterAttrs(AttributeSet Attrs, Type *Ty,
+      Check(Current.getLower().sge(Previous.getLower()),
+            "Attribute 'initializes' requires intervals in ascending order!",
+            V);
+      Check(Current.getLower().sge(Previous.getUpper()),
```
should this be sgt ?

haopliu replied:
Done!

jvoung commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,108 @@
+
+  void append(const ConstantRange &Range) {
+    assert(Range.getLower().slt(Range.getUpper()));
+    if (isFullSet())
```
assert that the bitwidth is the same for the new vs existing elements?

haopliu replied:
Done!

haopliu committed:
Update CRL about accessing the range of an EmptySet/FullSet

haopliu commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,108 @@
+  }
+
+  void append(int64_t Lower, int64_t Upper) {
+    append(ConstantRange(APInt(64, StringRef(std::to_string(Lower)), 10),
```
Done!

haopliu commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,108 @@
+    return !operator==(CRL);
+  }
+
+  /// Print out the bounds to a stream.
```
Done!

haopliu commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,108 @@
+
+  ConstantRangeList(int64_t Lower, int64_t Upper);
+
+  SmallVectorImpl<ConstantRange>::iterator begin() { return Ranges.begin(); }
```
Good point! Add assert to disallow this kind of accessing.

haopliu commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,108 @@
+  /// Get the bit width of this ConstantRangeList.
+  uint32_t getBitWidth() const { return Ranges[0].getBitWidth(); }
+
+  size_t size() const { return Ranges.size(); }
```
Documented in the class head!

aeubanks replied:
First sentence should be what this returns, second sentence clarifying Empty/FullSet.

don't abbreviate `CRL` in comment

haopliu replied:
Updated!

haopliu replied:
Updated the comment of `size()`.

jvoung commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,118 @@
+      return false;
+    for (size_t i = 0; i < size(); ++i) {
+      if (Ranges[i] != CRL.Ranges[i])
+        return false;
```
What is expected if one side is not `Sorted == true`? Should this be treated as a set?

There's also no read of the field `Sorted`.

How about enforcing that during "append" things remain sorted and then don't need the field?

haopliu replied:
Removed this field and enforced that in CRL::insert.

aeubanks replied:
`return Ranges == CRL.Ranges;`?

jvoung commented on llvm/lib/Bitcode/Reader/BitcodeReader.cpp:
```diff
@@ -2313,6 +2316,21 @@ Error BitcodeReader::parseAttributeGroupBlock() {
+            return error("Not a constant range list attribute");
+
+          ConstantRangeList CRL(64, false);
+          int RangeSize = Record[++i];
```
Maybe assert i + (RangeSize * 2) < e
(please double check if I got that right)

haopliu replied:
Done!

jvoung replied:
each loop iter up to `RangeSize` will bump `i` twice, so should it be * 2 ?

haopliu replied:
Oops, good catch! Done.

jvoung commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,44 @@
+
+ConstantRangeList::ConstantRangeList(int64_t Lower, int64_t Upper) {
+  Ranges.push_back(
+      ConstantRange(APInt(64, StringRef(std::to_string(Lower)), 10),
```
This one could also use the numeric APInt ctor instead of going through strings?

Is this ctor used by anything?

haopliu replied:
Good catch! A leftover useless ctor.

jvoung commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,118 @@
+    }
+    if (Range.getLower().slt(Ranges[size() - 1].getLower()))
+      Sorted = false;
+    Ranges.push_back(Range);
```
Related to the other question, should this be treated as a set? Handling of append when the new value overlaps with an existing element?

haopliu replied:
Make sense! Renamed the method to "insert" and enforced to be ordered and no overlapping.

haopliu committed:
Change CRL.insert to be ordered and no overlapping

jvoung commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,83 @@
+
+  ConstantRange RangeToInsert = Range;
+  SmallVector<ConstantRange, 2> ExistingRanges(Ranges.begin(), Ranges.end());
+  Ranges.clear();
```
Thanks for adjusting the append -> insert and keeping things sorted. I do realize that most current uses this will insert at the end, so the old "append" made sense in that way.

Is it possible to optimize this a bit more to avoid potential quadratic behavior in the common case of insert at end?

- instead of copying to ExistingRanges and clearing Ranges, keep as much of Ranges for Case 1, and only have potential copying for the tail where it becomes case 2 or 3?
- std::lower_bound might help find where case 2 and 3 start?

haopliu replied:
Sure, further optimized and early return for common cases. Please take another look.
It turns out that we need both "append" and "insert". Basically, use "insert" while inferring attr; use "append" in BitcodeReader/Assembler.

jvoung reviewed (commented):
Just a few small comments, otherwise LG to me. Not sure if other reviewers have comments. Thanks!

aeubanks reviewed (commented):
needs a test in `llvm/test/Bitcode/attributes.ll`, plus some verifier and assembler tests

see tests added in https://github.com/llvm/llvm-project/pull/84617

aeubanks commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,105 @@
+  /// Initialize a full or empty set for the specified bit width.
+  explicit ConstantRangeList(uint32_t BitWidth, bool isFullSet);
+
+  ConstantRangeList(int64_t Lower, int64_t Upper);
```
I'm not seeing a definition for this?

haopliu replied:
Removed!

aeubanks commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,105 @@
+
+public:
+  /// Initialize a full or empty set for the specified bit width.
+  explicit ConstantRangeList(uint32_t BitWidth, bool isFullSet);
```
if we're only using 64 bit widths, I'd remove the `BitWidth` param, especially since the other constructor seems to assume 64 bit width. if there's a need for more than just 64 in the future, we can add that then

haopliu replied:
Make sense! Removed the `BitWidth` parameter and set to 64 bits by default.

aeubanks commented on llvm/lib/IR/Attributes.cpp:
```diff
@@ -616,6 +661,23 @@ std::string Attribute::getAsString(bool InAttrGrp) const {
+    ConstantRangeList CRL = getValueAsConstantRangeList();
+    OS << "initializes(";
+    size_t i = 0;
+    for (auto &CR : CRL) {
```
use the recently introduced `interleaveComma`

haopliu replied:
Nice util! Applied `interleave` instead of `interleaveComma` to avoid the space after comma.

aeubanks commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,105 @@
+
+  /// Return true if this set contains all of the elements possible
+  /// for this data-type.
+  bool isFullSet() const { return Ranges.size() == 1 && Ranges[0].isFullSet(); }
```
we're never going to infer a full set right? (I mean theoretically we could on a function that immediately unwinds, but we could just use (min, max) in that case) so we shouldn't need to support `isFullSet`? would need to assert that any `ConstantRange` in `insert` is not a full set and also not an empty set

and with that we could change `isEmptySet` to `Ranges` being empty

haopliu replied:
Good point! Removed special `EmptySet` and `FullSet`. Now, empty ranges list means `Empty` and we don't support `FullSet`.

aeubanks commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,105 @@
+  /// For EmptySet or FullSet, the CRL size is 1 not 0.
+  size_t size() const { return Ranges.size(); }
+
+  /// Insert a range to Ranges. Keep the list ordered
```
this comment is mostly describing implementation details which I wouldn't put here.

haopliu replied:
Updated!

aeubanks commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,105 @@
+                         APInt(64, Upper, /*isSigned=*/true)));
+  }
+
+  /// Return true if this range is equal to another range.
```
nit: this is a list of ranges, not a range

haopliu replied:
Done!

haopliu committed:
Optimize CRL::insert for common cases

haopliu committed:
Update ConstantRangeList

haopliu committed:
Add unit tests for BitcodeReader/Verifier/Assembler

haopliu commented:
Addressed Arthur's and Jan's comments, and added unit tests for BitcodeReader/Verifier/Assembler!
Please take another look :-D
Gentle ping other reviewers. Thanks!

jvoung commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  // Slow insert.
+  SmallVector<ConstantRange, 2> ExistingRanges(Ranges.begin(), Ranges.end());
```
It seems like the ExistingRanges isn't using the part before LowerBound (below loop Iter = LowerBound and up). So is it possible to do something like this instead?
LowerBound = std::lower_bound(Ranges.begin(), ...)
ExistingTail(LowerBound, Ranges.end());
Ranges.erase(LowerBound, Ranges.end());
...
for (auto Iter = ExistingTail.begin(); ...) { ... }

something like that?

haopliu replied:
Done. More concise now :-D

jvoung reviewed (commented):
Thanks!

aeubanks commented on llvm/test/Verifier/initializes-attr.ll:
```diff
@@ -0,0 +1,31 @@
+
+; CHECK: Attribute 'initializes' requires interval lower less than upper
+; CHECK-NEXT: ptr @lower_greater_than_upper1
+define void @lower_greater_than_upper1(ptr initializes((4,0)) %a) {
```
nit: can we add a space after commas?

haopliu replied:
Sure, done!

haopliu committed:
Add a space after commas

haopliu commented:
Here is the new compile time cost for the entire stack of patches:
(Attr support with CRL + Attr inference + Apply in DSE).

https://llvm-compile-time-tracker.com/compare.php?from=07d7b9c255078edc6f04bd4e68416bdf3e8735ab&to=3d53c17c11aec6d71a35c5a80ef91764b23efb52&stat=instructions%3Au

See the detail in ["haopliu/perf/initialized-attr-CRL"](https://llvm-compile-time-tracker.com/index.php?config=Overview&stat=instructions%3Au&remote=haopliu).

nikic commented:
A quick question before I look into this in more detail: Why do we need the dedicated ConstantRangeList class? I'd have expected to see this passed around as `ArrayRef<ConstantRange>` and stored as `TrailingObjects<ConstantRange>`.

I see it has some helpers to insert ranges in sorted order, but that seems more like something for the inference implementation rather than the attribute implementation.

fhahn commented:
Would it be possible to include links to the follow-on patches, including inference + DSE?

haopliu commented:
"Why do we need the dedicated ConstantRangeList class? "
In the inference part, we need to insert ranges in sorted order, and union two sorted lists.
In the DSE part, we need to intersect two sorted lists (for example, intersect aliasing arguments' attribute lists). 
That's why I thought it sounds better to have a dedicated ConstantRangeList class. Otherwise, these util functions would be dispersed to multiple components with a lose control/code-review. 

Here is the inference patch: https://github.com/llvm/llvm-project/commit/e360135b22a64af497a9200e46172665e7935f14
Here is the patch to apply in DSE: https://github.com/llvm/llvm-project/commit/3d53c17c11aec6d71a35c5a80ef91764b23efb52

aeubanks commented on llvm/test/Assembler/initializes-attribute-invalid.ll:
```diff
@@ -0,0 +1,71 @@
+; RUN: split-file %s %t
+; RUN: not llvm-as < %s %t/outer_left_parenthesis.ll -o /dev/null 2>&1 | FileCheck %s --check-prefix=OUTER-LEFT
+; RUN: not llvm-as < %s %t/inner_left_parenthesis.ll -o /dev/null 2>&1 | FileCheck %s --check-prefix=INNER-LEFT
+; RUN: not llvm-as < %s %t/inner_right_parenthesis.ll -o /dev/null 2>&1 | FileCheck %s --check-prefix=OUTER-RIGHT
```
`--check-prefix=INNER-RIGHT`, although I guess they check the same thing :)

haopliu replied:
Ah, good catch! Done.

aeubanks commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,69 @@
+
+using namespace llvm;
+
+void ConstantRangeList::insert(const ConstantRange &NewRange) {
```
I wonder how often we insert a range that we already have in the list, then we could return immediately without doing any work

haopliu replied:
Added a common case that if `NewRange` exists in `Ranges`, early return!

aeubanks commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,69 @@
+      Ranges.back() = ConstantRange(NewLower, NewUpper);
+    }
+  }
+  return;
```
redundant `return;`

haopliu replied:
Removed!

aeubanks commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,69 @@
+
+void ConstantRangeList::print(raw_ostream &OS) const {
+  for (const auto &Range : Ranges)
+    Range.print(OS);
```
do we want to format this a little more nicely? like `((1, 2), (3, 4))`, as we print in textual IR

haopliu replied:
Sure, updated and call `CRL.print(OS)` in `Attributes.cpp` :-D

aeubanks reviewed (commented):
looks pretty good to me. I would like to take a look at the inference and DSE patches to make sure they look good at a high level before merging this one

haopliu committed:
Update CRL and unit tests

aeubanks commented on llvm/unittests/IR/ConstantRangeListTest.cpp:
```diff
@@ -0,0 +1,62 @@
+
+using ConstantRangeListTest = ::testing::Test;
+
+TEST_F(ConstantRangeListTest, Basics) {
```
can you add a test for (0,4) + (4,8)?

haopliu replied:
Nice catch! Changed one "slt" to "sle" in CRL::insert() and updated the unit test to cover this consecutive case.

haopliu committed:
Update CRL to merge consecutive ranges

aeubanks commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,73 @@
+    Ranges.insert(Ranges.begin(), NewRange);
+    return;
+  }
+  if (std::find(Ranges.begin(), Ranges.end(), NewRange) != Ranges.end()) {
```
we should be able to check just one element (instead of looking through all) below by looking at `LowerBound`?

haopliu replied:
Sure, handled the singleton case specially.

haopliu replied:
Oh, thanks for the reminder! Updated!

aeubanks commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,73 @@
+  }
+}
+
+void ConstantRangeList::print(raw_ostream &OS) const {
```
also could you add a `dump()` method? look around for `LLVM_DUMP_METHOD`, it usually just forwards to `print(errs())`, and is only defined in `#ifndef NDEBUG` builds

haopliu replied:
Done! Look like many dumps forward to `dbgs()`. Let me know if we prefer `errs()`.

haopliu committed:
Handle a common case in CRL::insert and add CRL::dump

haopliu committed:
Nit: change dbgs() to llvm::dbgs()

haopliu committed:
Change the attr impl to TrailingObjects<ConstantRange>

haopliu commented:
> A quick question before I look into this in more detail: Why do we need the dedicated ConstantRangeList class? I'd have expected to see this passed around as `ArrayRef<ConstantRange>` and stored as `TrailingObjects<ConstantRange>`.
> 
> I see it has some helpers to insert ranges in sorted order, but that seems more like something for the inference implementation rather than the attribute implementation.

Changed the attribute implementation to `TrailingObjects<ConstantRange>` + `ArrayRef<ConstantRange>`, and meanwhile kept the helper class ConstantRangeList for `intersectWith/unionWith/...`  in `FunctionAttrs.cpp` and `DSE.cpp`.

Please take another look. Thanks!

haopliu committed:
Update LLVMContextImpl.h

haopliu commented on llvm/lib/IR/LLVMContextImpl.h:
```diff
@@ -1565,6 +1566,8 @@ class LLVMContextImpl {
   UniqueStringSaver Saver{Alloc};
   SpecificBumpPtrAllocator<ConstantRangeAttributeImpl>
       ConstantRangeAttributeAlloc;
+  SpecificBumpPtrAllocator<ConstantRangeListAttributeImpl>
```
Yes, done!

haopliu commented on llvm/lib/IR/AttributeImpl.h:
```diff
@@ -222,6 +241,40 @@ class ConstantRangeAttributeImpl : public EnumAttributeImpl {
+      private TrailingObjects<ConstantRangeListAttributeImpl, ConstantRange> {
+  friend TrailingObjects;
+
+  unsigned ValSize;
```
Done.

haopliu commented:
Here is an updated compile time regression of the entire stack of patches: 
   this current RP + attribute inference with a dataflow approach ([53bd30](https://github.com/llvm/llvm-project/commit/53bd30bda61aac3b9ee098a5f8add0b21ef87d97)) + apply the attr in DSE ([d8d842](https://github.com/llvm/llvm-project/commit/d8d8420746ad988173adf5d71e9d51cd8d3975e7)).
https://llvm-compile-time-tracker.com/compare.php?from=07d7b9c255078edc6f04bd4e68416bdf3e8735ab&to=d8d8420746ad988173adf5d71e9d51cd8d3975e7&stat=instructions%3Au

aeubanks reviewed (approved):
lgtm with one last nit

I'd like some other approvals before landing this

for the followup patches, the inference patch generally lgtm (since I wrote a lot of it :) ), and the DSE patch in principle makes sense (the important thing is that it allows an instruction to be associated with multiple MemoryDefs), although I still need to understand it better. 0.07% compile time regression for interprocedural DSE seems acceptable IMO

jvoung commented on llvm/include/llvm/IR/Attributes.h:
```diff
@@ -267,6 +281,9 @@ class Attribute {
   ConstantRange getRange() const;
 
+  /// Returns the value of the initializes attribute.
+  ConstantRangeList getInitializes() const;
```
nit: Perhaps return ArrayRef<ConstantRange> to be consistent with getValueAsConstantRangeList(), then the caller can decide to make it a ConstantRangeList or not?

Also ArrayRef allows the caller to avoid a copy if it isn't needed.

haopliu replied:
Done!

jvoung reviewed (approved):
Nice work reducing the compile time from the initial version =)

haopliu committed:
Merge branch 'llvm:main' into dse-commit

haopliu committed:
Rebase to the latest main

haopliu committed:
Change getInitializes() to return ArrayRef<ConstantRange>

haopliu committed:
Remove llvm::

haopliu commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,80 @@
+
+#if !defined(NDEBUG) || defined(LLVM_ENABLE_DUMP)
+LLVM_DUMP_METHOD void ConstantRangeList::dump() const {
+  print(llvm::dbgs());
```
Done!

nikic commented on llvm/docs/LangRef.rst:
```diff
@@ -1621,6 +1621,26 @@ Currently, only the following parameter attributes are defined:
+    parameter. Other arbitrary accesses to the same memory via other pointers
+    are allowed.
+
+    The ``writable`` or ``dereferenceable`` attribute does not imply the
```
```suggestion
    The ``writable`` or ``dereferenceable`` attribute do not imply the
```

nikic commented on llvm/docs/LangRef.rst:
```diff
@@ -1621,6 +1621,26 @@ Currently, only the following parameter attributes are defined:
+    are allowed.
+
+    The ``writable`` or ``dereferenceable`` attribute does not imply the
+    ``initializes`` attribute. The ``initializes`` does not imply ``writeonly``
```
```suggestion
    ``initializes`` attribute. The ``initializes`` attribute does not imply ``writeonly``
```

haopliu replied:
Done!

nikic commented on llvm/docs/LangRef.rst:
```diff
@@ -1621,6 +1621,26 @@ Currently, only the following parameter attributes are defined:
+    since ``initializes`` allows reading from the pointer after writing.
+
+    This attribute is a list of constant ranges in ascending order with no
+    overlapping or consecutive list elements. ``LoN/HiN`` are 64-bit ints, and
```
```suggestion
    overlapping or consecutive list elements. ``LoN/HiN`` are 64-bit integers, and
```

Though I wonder whether the integers shouldn't match the pointer index type rather than being hardcoded to 64-bits...

haopliu replied:
Done!

nikic commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,86 @@
+//===- ConstantRangeList.h - A list of constant range -----------*- C++ -*-===//
```
```suggestion
//===- ConstantRangeList.h - A list of constant ranges -----------*- C++ -*-===//
```

haopliu replied:
Done!

nikic commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,86 @@
+//===----------------------------------------------------------------------===//
+//
+// Represent a list of signed ConstantRange and do NOT support wrap around the
+// end of the numeric range. Ranges in the list are ordered and no overlapping.
```
```suggestion
// end of the numeric range. Ranges in the list are ordered and not overlapping.
```
or "non-overlapping".

haopliu replied:
Done!

nikic commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,86 @@
+    return Ranges.end();
+  }
+  ConstantRange getRange(unsigned i) const {
+    assert(i < Ranges.size());
```
`operator[]` already has this assert.

haopliu replied:
Done!

nikic commented on llvm/lib/Bitcode/Reader/BitcodeReader.cpp:
```diff
@@ -2332,6 +2335,23 @@ Error BitcodeReader::parseAttributeGroupBlock() {
+            return error("Not a constant range list attribute");
+
+          SmallVector<ConstantRange, 2> Val;
+          int RangeSize = Record[++i];
```
```suggestion
          unsigned RangeSize = Record[++i];
```

haopliu replied:
Done!

nikic commented on llvm/lib/Bitcode/Reader/BitcodeReader.cpp:
```diff
@@ -2332,6 +2335,23 @@ Error BitcodeReader::parseAttributeGroupBlock() {
+
+          SmallVector<ConstantRange, 2> Val;
+          int RangeSize = Record[++i];
+          assert(i + 2 * RangeSize < e);
```
This can't be an assert, it should be an if with error return.

haopliu replied:
Ah, changed to an if with error return!

nikic commented on llvm/lib/Bitcode/Writer/BitcodeWriter.cpp:
```diff
@@ -943,11 +946,20 @@ void ModuleBitcodeWriter::writeAttributeGroupTable() {
+        Record.push_back(Val.size());
+        for (auto &CR : Val) {
+          emitSignedInt64(Record, CR.getLower().getSExtValue());
+          emitSignedInt64(Record, CR.getUpper().getSExtValue());
```
This hardcodes the 64-bit ranges in the bitcode format. We shouldn't do that to allow future uses of this attribute type with different width. This should use emitConstantRange() instead, maybe after a small refactoring to allow emitting the bit width separately (as we don't need to emit it for each range...)

aeubanks replied:
Ah I wasn't imagining that a future attribute might want to use ConstantRangeList, but that does seem like something we should support.

haopliu replied:
Changed to use `emitConstantRange()`.

nikic commented on llvm/lib/IR/Attributes.cpp:
```diff
@@ -191,6 +192,36 @@ Attribute Attribute::get(LLVMContext &Context, Attribute::AttrKind Kind,
+        ConstantRangeListAttributeImpl::totalSizeToAlloc(Val),
+        alignof(ConstantRangeListAttributeImpl));
+    PA = new (Mem) ConstantRangeListAttributeImpl(Kind, Val);
+    pImpl->AttrsSet.InsertNode(PA, InsertPoint);
```
I guess this ends up working fine right now where the ConstantRanges have <= 64 bits, but in general I think this may result in a memory leak due to the dtor not running.

haopliu replied:
How to call the dtor explicitly?

aeubanks replied:
hmm I'm not seeing where the existing ConstantRangeAttributeImpl's destructor gets called

aeubanks replied:
ah this should be using a separate `SpecificBumpPtrAllocator` like `ConstantRangeAttributeAlloc` so that we make sure to call destructors when the LLVMContext is destroyed

haopliu replied:
oh thanks for this info! Changed to `SpecificBumpPtrAllocator`. Please take another look :-)

nikic commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,80 @@
+// Represent a list of signed ConstantRange and do NOT support wrap around the
+// end of the numeric range. Ranges in the list are ordered and no overlapping.
+// Ranges should have the same bitwidth. Each range's lower should be less than
+// its upper.
```
Don't repeat the header comment here.

haopliu replied:
Done!

nikic commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,86 @@
+
+  /// Print out the ranges to a stream.
+  void print(raw_ostream &OS) const;
+  void dump() const;
```
Doesn't this also need the `#ifdef` checks?

haopliu replied:
Done!

nikic reviewed (commented):
FWIW, what I originally had in mind was to actually not have separate ConstantRange and ConstantRangeList attribute kinds, and only have the latter -- the existing ConstantRange kind would just be a list with a single element.

The upside of doing that is that we have less attribute kinds to deal with :) The downside is that the representation for the single-attribute case becomes less efficient (because we have to explicitly store the size 1).

aeubanks commented:
> FWIW, what I originally had in mind was to actually not have separate ConstantRange and ConstantRangeList attribute kinds, and only have the latter -- the existing ConstantRange kind would just be a list with a single element.
> 
> The upside of doing that is that we have less attribute kinds to deal with :) The downside is that the representation for the single-attribute case becomes less efficient (because we have to explicitly store the size 1).

The less efficient single-range seems ok if the single range attributes aren't too common and have too many unique ranges. It's hard to predict the future to know if this will be the case. Supporting a separate ConstantRangeList attribute doesn't seem too bad though, so I'm leaning toward keeping them separate.

haopliu committed:
Update LangRef and ConstantRangeList

haopliu committed:
Update BitCodeReader/Writer

aeubanks commented on llvm/lib/Bitcode/Writer/BitcodeWriter.cpp:
```diff
@@ -943,11 +947,20 @@ void ModuleBitcodeWriter::writeAttributeGroupTable() {
+        Record.push_back(8);
+        Record.push_back(getAttrKindEncoding(Attr.getKindAsEnum()));
+        ArrayRef<ConstantRange> Val = Attr.getValueAsConstantRangeList();
+        Record.push_back(Val.size());
```
this needs to emit the bit width once so we don't hardcode 64-bit ranges in ConstantRangeList, then read that back in in the bitcode reader and use that as the ConstantRange/APInt bit width

haopliu replied:
Thanks for the reminder. Done!

haopliu committed:
Encode bitwidth once for ConstantRangeList kind attr

aeubanks commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,73 @@
+  // Slow insert.
+  SmallVector<ConstantRange, 2> ExistingTail(LowerBound, Ranges.end());
+  Ranges.erase(LowerBound, Ranges.end());
+  // "sle" instead of "slt" to merge consecutive ranges.
```
the `"sle" instead of "slt"` portion of the comment isn't necessary

haopliu replied:
Removed this portion!

haopliu committed:
Use SpecificBumpPtrAllocator to allocate mem and make sure to call destructors

jvoung commented on llvm/lib/Bitcode/Reader/BitcodeReader.cpp:
```diff
@@ -2332,6 +2335,25 @@ Error BitcodeReader::parseAttributeGroupBlock() {
+          if (i + 2 * RangeSize >= e)
+            return error("Incomplete constant range list");
+          for (unsigned Idx = 0; Idx < RangeSize; ++Idx) {
+            int64_t Start = BitcodeReader::decodeSignRotatedValue(Record[++i]);
```
The `int64_t` as type would still assume BitWidth <= 64.
 
Can you factor out part of readConstantRange() and reuse that (the part after having read the BitWidth from the record)?

haopliu replied:
Refactored `readConstantRange` and reused it to avoid the bitwidth assumption. Thanks!

haopliu committed:
Refactor readConstantRange and update BitcodeReader

aeubanks commented:
@nikic any more concerns?

aeubanks commented:
ping @nikic

haopliu committed:
Merge branch 'main' into dse-commit

nikic commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,86 @@
+// Represent a list of signed ConstantRange and do NOT support wrap around the
+// end of the numeric range. Ranges in the list are ordered and not overlapping.
+// Ranges should have the same bitwidth. Each range's lower should be less than
+// its upper.
```
Something I noticed is that while ConstantRangeList maintains the "ordered and not overlapping" invariant through manipulations, it does not check that it initially holds. And indeed, LLParser will create such invalid ConstantRangeLists, which will later get rejected by the Verifier. Is this how it is *intended* to work out?

aeubanks replied:
it is convenient that the verifier can check for sorted `ConstantRangeList`s without asserting on construction, but perhaps it is better to move the checks to both LLParser and bitcode parser

a static method that returns a `std::optional<ConstantRangeList>` depending on if the input is sorted or not, and error checking in the LLParser/bitcode parser would be good. then we can assert that the input is sorted in the constructor.

(or just automatically sort ConstantRangeLists on construction, but IMO the checks can be useful for finding bugs)

haopliu replied:
Added `static std::optional<ConstantRangeList> getConstantRangeList()` and checked the optional return in LLParse and bitcode parser.

nikic commented on llvm/lib/Bitcode/Reader/BitcodeReader.cpp:
```diff
@@ -2337,12 +2348,35 @@ Error BitcodeReader::parseAttributeGroupBlock() {
+
+          SmallVector<ConstantRange, 2> Val;
+          unsigned RangeSize = Record[++i];
+          unsigned BitWidth = Record[++i];
```
Missing bounds check before these accesses.

haopliu replied:
Done!

nikic commented on llvm/lib/Bitcode/Reader/BitcodeReader.cpp:
```diff
@@ -2337,12 +2348,35 @@ Error BitcodeReader::parseAttributeGroupBlock() {
+          if (i + 2 * RangeSize >= e)
+            return error("Incomplete constant range list");
+          for (unsigned Idx = 0; Idx < RangeSize; ++Idx) {
+            i++;
```
I don't really get why this i++ and i-- dance here is necessary. I think maybe you want just one ++i *before* the loop?

haopliu replied:
Moved to before the loop and changed the Record access to "i++" rather than "++i", the same way as the "Range" attribute.

nikic commented on llvm/lib/Bitcode/Writer/BitcodeWriter.cpp:
```diff
@@ -953,11 +957,21 @@ void ModuleBitcodeWriter::writeAttributeGroupTable() {
+        Record.push_back(Val[0].getBitWidth());
+        for (auto &CR : Val) {
+          emitConstantRange(Record, CR, /*EmitBitWidth=*/false);
+        }
```
Remove braces.

haopliu replied:
Done!

nikic commented on llvm/lib/IR/AttributeImpl.h:
```diff
@@ -222,6 +241,37 @@ class ConstantRangeAttributeImpl : public EnumAttributeImpl {
+      : EnumAttributeImpl(ConstantRangeListAttrEntry, Kind), Size(Val.size()) {
+    ConstantRange *TrailingCR = getTrailingObjects<ConstantRange>();
+    assert(Size > 0);
+    unsigned BitWidth = Val.front().getLower().getBitWidth();
```
```suggestion
    unsigned BitWidth = Val.front().getBitWidth();
```

haopliu replied:
Removed this code snippet.

nikic commented on llvm/lib/IR/AttributeImpl.h:
```diff
@@ -222,6 +241,37 @@ class ConstantRangeAttributeImpl : public EnumAttributeImpl {
+    assert(Size > 0);
+    unsigned BitWidth = Val.front().getLower().getBitWidth();
+    for (unsigned I = 0; I != Size; ++I) {
+      assert(BitWidth == Val[I].getLower().getBitWidth());
```
```suggestion
      assert(BitWidth == Val[I].getBitWidth());
```

haopliu replied:
Removed this code snippet.

nikic commented on llvm/lib/IR/AttributeImpl.h:
```diff
@@ -222,6 +241,37 @@ class ConstantRangeAttributeImpl : public EnumAttributeImpl {
+    unsigned BitWidth = Val.front().getLower().getBitWidth();
+    for (unsigned I = 0; I != Size; ++I) {
+      assert(BitWidth == Val[I].getLower().getBitWidth());
+      new (&TrailingCR[I]) ConstantRange(BitWidth, false);
```
Can't you use `uninitialized_copy` instead of performing a dummy initialization first?

haopliu replied:
Removed this code snippet.

haopliu replied:
Changed to `uninitialized_copy`!

nikic commented on llvm/lib/IR/Attributes.cpp:
```diff
@@ -191,6 +192,35 @@ Attribute Attribute::get(LLVMContext &Context, Attribute::AttrKind Kind,
+    // If we didn't find any existing attributes of the same shape then create a
+    // new one and insert it.
+    PA = new (pImpl->ConstantRangeListAttributeAlloc.Allocate(
+        ConstantRangeListAttributeImpl::totalSizeToAlloc(Val)))
```
Heh, this does something completely different from what you think it does. It allocates an *array* of `sizeof(ConstantRangeListAttributeImpl) * totalSizeToAlloc(Val)`.

It's not possible to use SpecificBumpPtrAllocator with a dynamically sized class, it can only be used to allocate objects of fixed size.

aeubanks replied:
if we have a variable number of `ConstantRange`s in a `ConstantRangeListAttributeImpl` and they need a destructor, I don't see a nice way of using trailing objects and bump allocators. I think we should go back to `ConstantRangeListAttributeImpl` containing a `ConstantRangeList` and using `SpecificBumpPtrAllocator<ConstantRangeListAttributeImpl>` so that we allocate the right amount of memory and properly call the destructor. most of the time we'll be in the non-allocating version of the `SmallVector<ConstantRange>` anyway

haopliu replied:
Changed the attr implementation back to ConstantRangeList.

haopliu committed:
Change the attr imple back to ConstantRangeList

haopliu commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,86 @@
+  bool empty() const { return Ranges.empty(); }
+
+  /// Get the bit width of this ConstantRangeList.
+  uint32_t getBitWidth() const { return 64; }
```
Yeah, relaxing later on demand is the plan!

haopliu committed:
Add getConstantRangeList unittest

haopliu commented:
Thanks for the comments, Nikita and Arthur! @nikic please take another look.

aeubanks commented on llvm/include/llvm/IR/ConstantRangeList.h:
```diff
@@ -0,0 +1,87 @@
+public:
+  ConstantRangeList() = default;
+  ConstantRangeList(ArrayRef<ConstantRange> RangesRef) {
+    for (const ConstantRange &R : RangesRef) {
```
add asserts that these are non-overlapping and increasing

haopliu replied:
Done!

aeubanks commented on llvm/lib/Bitcode/Reader/BitcodeReader.cpp:
```diff
@@ -858,6 +859,14 @@ class BitcodeReader : public BitcodeReaderBase, public GVMaterializer {
 
+  Expected<ConstantRange>
+  readBitWidthAndConstantRange(ArrayRef<uint64_t> Record, unsigned &OpNum) {
+    if (Record.size() - OpNum < 3)
```
I'd check `1` instead of `3` and let `readConstantRange` deal with its own reading error

haopliu replied:
Done.

aeubanks commented on llvm/lib/IR/AttributeImpl.h:
```diff
@@ -124,6 +133,16 @@ class AttributeImpl : public FoldingSetNode {
+  static void Profile(FoldingSetNodeID &ID, Attribute::AttrKind Kind,
+                      const ConstantRangeList &CRL) {
+    ID.AddInteger(Kind);
+    ID.AddInteger(CRL.size());
```
should also add bit width here now

haopliu replied:
Ah, nice catch. Done!

aeubanks reviewed (commented):
@nikic ping

haopliu committed:
Add Bitwidth in FoldingSetNodeId

haopliu committed:
Merge branch 'main' into dse-commit

jvoung commented on llvm/lib/Bitcode/Reader/BitcodeReader.cpp:
```diff
@@ -2350,12 +2361,42 @@ Error BitcodeReader::parseAttributeGroupBlock() {
+            return error("Too few records for constant range list");
+          unsigned RangeSize = Record[i++];
+          unsigned BitWidth = Record[i++];
+          if (i + 2 * RangeSize > e)
```
I think you can remove this hoisted `i + 2 * RangeSize > e` check, since it's not always 2 elements per RangeSize anymore and readConstantRange is doing the checks in the loop.

haopliu replied:
Ah, done!

haopliu committed:
Clean up a redundant check in BitcodeReader

aeubanks commented:
I think @nikic's concerns have been addressed and there are no major objections to this patch going in. We can address any post-commit feedback with followup PRs, but I'd like to get this in so we can start reviewing the DSE + inference PRs.

nikic commented:
Why did this change back to storing ConstantRangeList in the Attribute rather than the efficient TrailingObjects / ArrayRef representation? If this was just because of the memory leak issue, I think an easy to solve that would have been to a) use the normal Alloc for allocation (instead of a separate SpecificBumpPtrAllocator) and b) add a vector to which all allocated pointers are added. Then in the LLVMContext dtor call the dtors of everything in the vector.

haopliu committed:
Change attr impl to normal Alloc w/ a vector and call dtor for each attr in LLVMContext dtor

haopliu committed:
nit: use uninitialized_copy in the ConstantRangeListAttributeImpl ctor

haopliu commented:
Good point. Changed the attr impl to use normal Alloc w/ a vector and explicitly call each attr's dtor in LLVMContext dtor. Please take another look. Thanks!

nikic commented on llvm/lib/IR/AttributeImpl.h:
```diff
@@ -124,6 +133,17 @@ class AttributeImpl : public FoldingSetNode {
+                      ArrayRef<ConstantRange> Val) {
+    ID.AddInteger(Kind);
+    ID.AddInteger(Val.size());
+    ID.AddInteger(Val[0].getBitWidth());
```
This is not necessary, because APInt::Profile already includes it.

haopliu replied:
Ah, done!

nikic commented on llvm/lib/IR/Attributes.cpp:
```diff
@@ -191,6 +192,44 @@ Attribute Attribute::get(LLVMContext &Context, Attribute::AttrKind Kind,
+  FoldingSetNodeID ID;
+  ID.AddInteger(Kind);
+  ID.AddInteger(Val.size());
+  ID.AddInteger(Val[0].getBitWidth());
```
Same here.

haopliu replied:
Done.

nikic commented on llvm/lib/IR/Attributes.cpp:
```diff
@@ -191,6 +192,44 @@ Attribute Attribute::get(LLVMContext &Context, Attribute::AttrKind Kind,
+  if (!PA) {
+    // If we didn't find any existing attributes of the same shape then create a
+    // new one and insert it.
+    // ConstantRange is a dynamically sized class and cannot use
```
```suggestion
    // ConstantRangeListAttributeImpl is a dynamically sized class and cannot use
```
ConstantRange itself is not dynamically sized.

haopliu replied:
Updated!

nikic commented on llvm/lib/IR/Attributes.cpp:
```diff
@@ -191,6 +192,44 @@ Attribute Attribute::get(LLVMContext &Context, Attribute::AttrKind Kind,
+    // new one and insert it.
+    // ConstantRange is a dynamically sized class and cannot use
+    // SpecificBumpPtrAllocator. Instead, we use normal Alloc for allocation
+    // for allocation and record the allocated pointer in
```
Duplicate "for allocation".

haopliu replied:
Done.

nikic commented on llvm/lib/IR/Attributes.cpp:
```diff
@@ -191,6 +192,44 @@ Attribute Attribute::get(LLVMContext &Context, Attribute::AttrKind Kind,
+    // SpecificBumpPtrAllocator. Instead, we use normal Alloc for allocation
+    // for allocation and record the allocated pointer in
+    // `ConstantRangeListAttributes`. LLVMContext destructor will call the
+    // destuctor of the allocated pointer explicitly.
```
```suggestion
    // destructor of the allocated pointer explicitly.
```

haopliu replied:
Done.

nikic commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,97 @@
+    auto CurRange = RangesRef[i];
+    auto PreRange = RangesRef[i - 1];
+    if (CurRange.getLower().sge(CurRange.getUpper()) ||
+        CurRange.getLower().sle(PreRange.getLower()) ||
+        CurRange.getLower().sle(PreRange.getUpper()))
```
```suggestion
        CurRange.getLower().sle(PreRange.getUpper()))
```
Should be sufficient?

haopliu replied:
Removed. `CurRange.getLower().sle(PreRange.getLower())` is transitively checked :)

nikic commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  auto LowerBound =
+      std::lower_bound(Ranges.begin(), Ranges.end(), NewRange,
```
```suggestion
      lower_bound(Ranges, NewRange,
```

haopliu replied:
Done

nikic commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,97 @@
+                       [](const ConstantRange &a, const ConstantRange &b) {
+                         return a.getLower().slt(b.getLower());
+                       });
+  if (LowerBound != Ranges.end() && *LowerBound == NewRange)
```
Instead of checking strict equality here, could check whether LowerBound contains NewRange instead? That would handle more cases via fast path.

haopliu replied:
Yes, changed to check whether LowerBound contains NewRange.

nikic commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -2059,6 +2060,27 @@ void Verifier::verifyParameterAttrs(AttributeSet Attrs, Type *Ty,
+          V);
+
+    Check(Inits[0].getLower().slt(Inits[0].getUpper()),
+          "Attribute 'initializes' requires interval lower less than upper", V);
```
Reuse isOrderedRanges here?

haopliu replied:
Done!

nikic reviewed (commented):
A few more nits, but mostly this looks good to me.

haopliu committed:
Remove redundant BitWidth in the attr Profile as APInt::Profile already include it.

nikic commented on llvm/lib/IR/ConstantRangeList.cpp:
```diff
@@ -0,0 +1,97 @@
+        return a.getLower().slt(b.getLower());
+      });
+  if (LowerBound != Ranges.end() &&
+      LowerBound->getLower().eq(NewRange.getLower()) &&
+      LowerBound->getUpper().sge(NewRange.getUpper()))
```
```suggestion
      LowerBound->contains(NewRange))
```

haopliu replied:
Ah, didn't notice `ConstantRange::contains()`. Nice!

nikic reviewed (approved):
LGTM

haopliu committed:
Use ConstantRange::contains() in ConstantRangeList::insert()
